File: src/embroidermodder/statusbar.py
#!/usr/bin/env python3

r"""
    Embroidermodder 2.

    -----

    Copyright 2013-2022 The Embroidermodder Team
    Embroidermodder 2 is Open Source Software.
    See LICENSE for licensing terms.

    -----

    Another attempt at a graphical user interface that runs on
    lots of machines without a complex build or fragile dependencies.

    This is a translation of some of the ideas we came up with for other
    attempts.
"""

import tkinter as tk


class StatusBarButton():
    r"""
    StatusBarButton(buttonText, mw, statbar, parent = 0)
    StatusBar*  statusbar
    def contextMenuEvent(QContextMenuEvent *event = 0)
    """
    def __init__(self, buttonText, mw, statbar, parent):
        statusbar = statbar

        this.set_object_name("StatusBarButton" + buttonText)

        this.setText(buttonText)
        this.setAutoRaise(1)
        this.setCheckable(1)

        if self.object_name == "SNAP":
            connect(this, SIGNAL(toggled(int)), this, SLOT(toggleSnap(int)))
        elif self.object_name == "GRID":
            connect(this, SIGNAL(toggled(int)), this, SLOT(toggleGrid(int)))
        elif self.object_name == "RULER":
            connect(this, SIGNAL(toggled(int)), this, SLOT(toggleRuler(int)))
        elif self.object_name == "ORTHO":
            connect(this, SIGNAL(toggled(int)), this, SLOT(toggleOrtho(int)))
        elif self.object_name == "POLAR":
            connect(this, SIGNAL(toggled(int)), this, SLOT(togglePolar(int)))
        elif self.object_name== "QSNAP":
            connect(this, SIGNAL(toggled(int)), this, SLOT(toggleQSnap(int)))
        elif self.object_name== "QTRACK":
            connect(this, SIGNAL(toggled(int)), this, SLOT(toggleQTrack(int)))
        elif self.object_name== "LWT":
            connect(this, SIGNAL(toggled(int)), this, SLOT(toggleLwt(int)))

    def contextMenuEvent(self, event):
        QApplication_setOverrideCursor(Qt_ArrowCursor)
        menu_ = QMenu(this)
        if self.object_name== "SNAP":
            settingsSnapAction = Action(loadIcon(gridsnapsettings_xpm), "Settings...", menu_)
            connect(settingsSnapAction, SIGNAL(triggered()), this, SLOT(settingsSnap()))
            menu_.addAction(settingsSnapAction)

        elif self.object_name== "GRID":
            settingsGridAction = Action(loadIcon(gridsettings_xpm), "Settings...", menu_)
            connect(settingsGridAction, SIGNAL(triggered()), this, SLOT(settingsGrid()))
            menu_.addAction(settingsGridAction)

        elif self.object_name== "RULER":
            settingsRulerAction = Action( Icon("icons/rulersettings.png"), "Settings...", menu_)
            connect(settingsRulerAction, SIGNAL(triggered()), this, SLOT(settingsRuler()))
            menu_.addAction(settingsRulerAction)

        elif self.object_name== "ORTHO":
            settingsOrthoAction = Action( Icon("icons/orthosettings.png"), "Settings...", menu_)
            connect(settingsOrthoAction, SIGNAL(triggered()), this, SLOT(settingsOrtho()))
            menu_.addAction(settingsOrthoAction)

        elif self.object_name== "POLAR":
            settingsPolarAction = Action( Icon("icons/polarsettings.png"), "Settings...", menu_)
            connect(settingsPolarAction, SIGNAL(triggered()), this, SLOT(settingsPolar()))
            menu_.addAction(settingsPolarAction)

        elif self.object_name== "QSNAP":
            settingsQSnapAction = Action( Icon("icons/qsnapsettings.png"), "Settings...", menu_)
            connect(settingsQSnapAction, SIGNAL(triggered()), this, SLOT(settingsQSnap()))
            menu_.addAction(settingsQSnapAction)

        elif self.object_name== "QTRACK":
            settingsQTrackAction = Action( Icon("icons/qtracksettings.png"), "Settings...", menu_)
            connect(settingsQTrackAction, SIGNAL(triggered()), this, SLOT(settingsQTrack()))
            menu_.addAction(settingsQTrackAction)

        elif self.object_name== "LWT":
            gview = main_win.active_view()
            if gview:
                enable_real_action = Action(Icon("icons/realrender.png"), "RealRender On", menu_)
                enable_real_action.setEnabled(!gview.isRealEnabled())
                connect(enable_real_action, SIGNAL(triggered()), this, SLOT(enableReal()))
                menu_.addAction(enable_real_action)

                disable_real_action = Action(Icon("icons/realrender.png"), "RealRender Off", menu_)
                disable_real_action.setEnabled(gview.isRealEnabled())
                connect(disable_real_action, SIGNAL(triggered()), this, SLOT(disableReal()))
                menu_.addAction(disable_real_action)

            settingsLwtAction = Action(loadIcon(lineweightsettings_xpm), "Settings...", menu_)
            connect(settingsLwtAction, SIGNAL(triggered()), this, SLOT(settingsLwt()))
            menu_.addAction(settingsLwtAction)

        menu_.exec(event.globalPos())
        restoreOverrideCursor()
        statusbar.clearMessage()

class StatusBar():
    """
    StatusBar(MainWindow* mw, QWidget* parent = 0)
    def setMouseCoord(x, y)
    """
    def __init__(self, mw, parent):
        " . "
        this.set_object_name("StatusBar")

        for i in range(N_STATUS):
            status_bar[i] = StatusBarButton(status_bar_label[i], main_win, this, this)

        status_bar_mouse_coord = tk.Label(this)

        status_bar_mouse_coord.set_minimum_width(300)
        # Must fit this text always
        status_bar_mouse_coord.set_maximum_width(300)
        # "+1.2345E+99, +1.2345E+99, +1.2345E+99"

        this.add_widget(status_bar_mouse_coord)
        for i in range(N_STATUS):
            this.add_widget(status_bar[i])

    def set_mouse_coord(x, y):
        " . "
        # TODO: set format from settings (Architectural, Decimal, Engineering, Fractional, Scientific)

        # Decimal
        status_bar_mouse_coord.setText("".setNum(x, 'F', 4) + ", "
            + "".setNum(y, 'F', 4));
            #TODO: use precision from unit settings

        # Scientific
        # status_bar_mouse_coord.setText("".setNum(x, 'E', 4)
            + ", " + "".setNum(y, 'E', 4));
        # TODO: use precision from unit settings

File: src/status_bar.c
/* 
 *  Embroidermodder 2.
 *
 *  ------------------------------------------------------------
 *
 *  Copyright 2013-2022 The Embroidermodder Team
 *  Embroidermodder 2 is Open Source Software.
 *  See LICENSE for licensing terms.
 *
 *  ------------------------------------------------------------
 *
 *  Use Python's PEP7 style guide.
 *      https://peps.python.org/pep-0007/
 *
 *  Status bar management.
 */

#include "em2.h"

void
status_bar_update(int status_bar)
{

}


/*
void status_bar_context_menu_event(
    WindowTab *status_bar, SDL_Event *event)
{
    setOverrideCursor(Qt-ArrowCursor)
    menu- = QMenu(this)
    switch (object-name) {
    case SNAP:
        settingsSnapAction = Action(load-icon(gridsnapsettings-xpm), "Settings...", menu-)
        connect(settingsSnapAction, SIGNAL(triggered()), this, SLOT(settingsSnap()))
        menu-.addAction(settingsSnapAction)
        break

    case GRID:
        settingsGridAction = Action(load-icon(gridsettings-xpm), "Settings...", menu-)
        connect(settingsGridAction, SIGNAL(triggered()), this, SLOT(settingsGrid()))
        menu-.addAction(settingsGridAction)
        break

    case RULER:
        settingsRulerAction = Action( Icon("icons/rulersettings.png"), "Settings...", menu-)
        connect(settingsRulerAction, SIGNAL(triggered()), this, SLOT(settingsRuler()))
        menu-.addAction(settingsRulerAction)
        break

    case ORTHO:
        settingsOrthoAction = Action( Icon("icons/orthosettings.png"), "Settings...", menu-)
        connect(settingsOrthoAction, SIGNAL(triggered()), this, SLOT(settingsOrtho()))
        menu-.addAction(settingsOrthoAction)
        break

    case POLAR:
        settingsPolarAction = Action( Icon("icons/polarsettings.png"), "Settings...", menu-)
        connect(settingsPolarAction, SIGNAL(triggered()), this, SLOT(settingsPolar()))
        menu-.addAction(settingsPolarAction)
        break

    case QSNAP:
        settingsQSnapAction = Action( Icon("icons/qsnapsettings.png"), "Settings...", menu-)
        connect(settingsQSnapAction, SIGNAL(triggered()), this, SLOT(settingsQSnap()))
        menu-.addAction(settingsQSnapAction)

    case QTRACK:
        settingsQTrackAction = Action( Icon("icons/qtracksettings.png"), "Settings...", menu-)
        connect(settingsQTrackAction, SIGNAL(triggered()), this, SLOT(settingsQTrack()))
        menu-.addAction(settingsQTrackAction)
        break

    case LWT:
        gview = main-win.active-view()
        if (gview) {
            enable-real-action = Action(Icon("icons/realrender.png"), "RealRender On", menu-)
            enable-real-action.setEnabled(!gview.isRealEnabled())
            connect(enable-real-action, SIGNAL(triggered()), this, SLOT(enableReal()))
            menu-.addAction(enable-real-action)

            disable-real-action = Action(Icon("icons/realrender.png"), "RealRender Off", menu-)
            disable-real-action.setEnabled(gview.isRealEnabled())
            connect(disable-real-action, SIGNAL(triggered()), this, SLOT(disableReal()))
            menu-.addAction(disable-real-action)
        }

        settingsLwtAction = Action(load-icon(lineweightsettings-xpm), "Settings...", menu-)
        connect(settingsLwtAction, SIGNAL(triggered()), this, SLOT(settingsLwt()))
        menu-.addAction(settingsLwtAction)
        break

    default:
        break
    }

    menu-.exec(event.globalPos());
    restore_override_cursor();
    statusbar.clearMessage();
}
*/

/* TODO: set format from settings.
 */
void
status_bar_mouse_coord(WindowTab *tab, int x, int y)
{
    printf("%d %d\n", x, y);
    /*
    if (tab->number-mode == ARCHITECTURAL) {
        return;
    }
    if (tab->number-mode == ENGINEERING) {
        return;
    }
    if (tab->number-mode == FRACTIONAL) {
        return;
    }
    if (tab->number-mode == SCIENTIFIC) { */
        /* status_bar-mouse-coord.setText("".setNum(x, 'E', 4)
            + ", " + "".setNum(y, 'E', 4))
        # TODO: use precision from unit settings */ /*
        return;
    } */
    /* Else decimal */
    /*
     * status_bar-mouse-coord.setText("".setNum(x, 'F', 4) + ", "
     * + "".setNum(y, 'F', 4))
     */ /* TODO: use precision from unit settings */
}

/* class StatusBar() */
/*
 * StatusBar( QWidget* parent = 0)
 * def setMouseCoord(x, y)
 */
void
status_bar_init(void)
{
    /*
    (self, buttonText, mw, statbar, parent);
    statusbar = statbar;

    this.set_object_name("StatusBarButton" + buttonText);

    this.setText(buttonText);
    this.setAutoRaise(1);
    this.setCheckable(1);

    if (object-name >= 0 && object-name < N-STATUS) {
        status[object-name] = !status[object-name];
    }
    */
    /*
    int i;
    def --init--(self, mw, parent)
    this.set_object_name("StatusBar");

    for (i=0; i<N_STATUS; i++) {
        status_bar[i] = StatusBarButton(status_bar-label[i], main-win, this, this);
    }

    status_bar-mouse-coord = tk.Label(this)

    status_bar-mouse-coord.set_minimum-width(300)
    # Must fit this text always
    status_bar-mouse-coord.set_maximum-width(300)
    # "+1.2345E+99, +1.2345E+99, +1.2345E+99"

    this, add_widget(status_bar-mouse-coord)
    for i in range(N-STATUS) {
        this, add_widget(status_bar[i])
    }
    */
}

